README.rst: extended README with usage section (#23)
RIOT-OS/riotctrl: README.rst

      lint: commands succeeded
      flake8: commands succeeded
      congratulations :)

Usage
-----

RIOTCtrl provides a python object abstraction of a RIOT device. It's meant
as a starting point for any serial abstraction on which higher level
abstractions (like a shell) can be built.

```python
from riotctrl.ctrl import RIOTCtrl

env = {'BOARD': 'native'}
# if not running from the application directory the a path must be provided
ctrl = RIOTCtrl(env=env, application_directory='.')
# flash the application
ctrl.make_run(['flash'])
# run the terminal through a contextmanager
with ctrl.run_term():
    ctrl.term.expect('>')       # wait for shell to start
    ctrl.term.sendline("help")  # send the help command
    ctrl.term.expect('>')       # wait for the command result to finnish
    print(ctrl.term.before)     # print the command result
# run without a contextmanager
ctrl.start_term()               # start a serial terminal
ctrl.term.sendline("help")      # send the help command
ctrl.term.expect('>')           # wait for the command result to finnish
print(ctrl.term.before)         # print the command result
ctrl.stop_term()                # close the terminal

> ```

Reviewer: This (and in other places) needs to be restructuredText, not Markdown. Currently, the output [is garbled](https://github.com/RIOT-OS/riotctrl/blob/7421debeb00cfc44b657dc9a7821fa58514a8ecb/README.rst#usage). You can use tools like pandoc to convert the formats.

Author: converted, thanks!

      lint: commands succeeded
      flake8: commands succeeded
      congratulations :)

Usage
-----

> RIOTCtrl provides a python object abstraction of a RIOT device. It’s

Reviewer: ```suggestion
RIOTCtrl provides a python object abstraction of a RIOT device. It's
```

meant as a starting point for any serial abstraction on which higher
level abstractions (like a shell) can be built.

.. code:: python

   from riotctrl.ctrl import RIOTCtrl

   env = {'BOARD': 'native'}
   # if not running from the application directory the a path must be provided
   ctrl = RIOTCtrl(env=env, application_directory='.')
   # flash the application
   ctrl.make_run(['flash'])
   # run the terminal through a contextmanager
   with ctrl.run_term():
       ctrl.term.expect('>')       # wait for shell to start
       ctrl.term.sendline("help")  # send the help command
       ctrl.term.expect('>')       # wait for the command result to finnish
       print(ctrl.term.before)     # print the command result
   # run without a contextmanager
   ctrl.start_term()               # start a serial terminal
   ctrl.term.sendline("help")      # send the help command
   ctrl.term.expect('>')           # wait for the command result to finnish
   print(ctrl.term.before)         # print the command result
   ctrl.stop_term()                # close the terminal

> Creating a RIOTCtrl object is done via environments. If empty then all

Reviewer: ```suggestion
Creating a RIOTCtrl object is done via environment variables. If empty then all
```
?

configuration will come from the target application makefile. But any

> Make environment variable can be overridden, for example setting

Reviewer: ```suggestion
environment variable can be overridden, for example setting
```

They are not just restricted to Make ;-)

``BOARD`` to a target ``BOARD`` which is not the default for that
application.

Any make target used on RIOT devices can be used on the abstraction

> like: ``make flash`` => ``ctrl.make_run(['flash'])``.

Reviewer: "on the abstraction like"? I think there is a word missing here.

Reviewer: Since this is the first mention of `make_run` I wouldn't replace it here, but maybe also mention `ctrl.flash()` here. Something like e.g.

```suggestion
Any make target used on RIOT devices can be used on the abstraction
like: ``make flash`` => ``ctrl.make_run(['flash'])``. Some make
targets are even further abstracted, e.g. ``ctrl.make_run(['flash'])``
can also be executed using the ``ctrl.flash()`` method.
```

``ctrl.start_term()`` (``make term``\ ’s alter ego) by default spawns a

> pexpect child application. From there interactions with the application

Reviewer: Maybe link `pexpect` here, for those unaware about it.

Author: Done

under use can be atomized. In the example below the output of the
``"help"`` command is captured:

ShellInteractions
~~~~~~~~~~~~~~~~~

> RIOTCtrl provides a minimal extensions by using pexpect replwrap: “[A]

Reviewer: Same here `pexpect replwrap` shoud be referenced.

Generic wrapper for read-eval-print-loops, a.k.a. interactive shells”.

> This implements the perfect wrapper for RIOT shell commands since it

Reviewer: Is the wrapper so perfect? ;-)

Author: would a "nice wrapper" be ok?

Reviewer: Yes :-).

will wait for a command to finish before returning its output.

RIOT already provides a ``ShellInteraction`` for the ``"help"`` command,

> it can be imported as ``from riotctrl_shell.sys import Help``.

Author: Done, hope thats good

The previous example can be re-written using ``ShellInteraction``:

.. code:: python

   from riotctrl.ctrl import RIOTCtrl
   from riotctrl.shell import ShellInteraction

   env = {'BOARD': 'native'}
   # if not running from the application directory the a path must be provided
   ctrl = RIOTCtrl(env=env, application_directory='.')
   # flash the application
   ctrl.flash()                     # alias for ctrl.make_run(['flash'])
   # shell interaction instance
   shell = ShellInteraction(ctrl)
   shell.start_term()               # start a serial terminal
   print(shell.cmd("help"))         # print the command result
   shell.stop_term()                # close the terminal

or using the already provided ``Help`` ``ShellInteraction``:

.. code:: python

   class Help(ShellInteraction):
       """Help ShellInteraction"""
       @ShellInteraction.check_term
       def help(self, timeout=-1, async_=False):
           """Sends the reboot command via the terminal"""

>            return self.cmd("help", timeout, async_)

Reviewer: That the `Help` shell interaction is printed here, confused me at first. Is this required to be copy-pasted in the code. No => maybe remove it.

Author: Linked instead

.. code:: python

   from riotctrl.ctrl import RIOTCtrl
   from riotctrl_shell.sys import Help

   env = {'BOARD': 'native'}
   # if not running from the application directory the a path must be provided
   ctrl = RIOTCtrl(env=env, application_directory='.')
   # flash the application
   ctrl.flash()                     # alias for ctrl.make_run(['flash'])
   # shell interaction instance, Help uses the @ShellInteraction.check_term
   # decorator, it will start the terminal if its not yet running, and close
   # it after the command ends

>    shell = Help(ctrl)              #

Reviewer: Comment missing?

   print(shell.help())             # print the command result

> Writing SAUL ShellInteraction

Reviewer: This section and the next could out-date pretty fast, as soon as there is a SAUL shellinteraction in upstream... Maybe do this on the basis of a shell command defined in this text. For instance, "Given a main file 'main.c' \<print main.c with some example shell command>, we can write the following ShellInteraction extension.

Author: Does its really matter? I'm giving the commands here, with the expected output, even if output changes in the future the example would remain valid.

Reviewer: Would just be more future-proof IMHO, to document this independent from RIOT-OS/RIOT. So yes, it matters.

Author: Done

   #0  SENSE_XX    x1
   #1  SENSE_XX    x2

Basically four commands a shell interaction for listing devices can be
written as:

.. code:: python

   from riotctrl.shell import ShellInteraction

   class SaulShell(ShellInteraction):
       @ShellInteraction.check_term
       def saul_read(self, timeout=-1, async_=False):
           return self.cmd("saul", timeout, async_)

Since all commands will share the same start: ``saul``, we can write
base same all on a basic ``saul`` cmd with extra args:

.. code:: python

   from riotctrl.shell import ShellInteraction

   class SaulShell(ShellInteraction):
       @ShellInteraction.check_term
       def saul_cmd(self, args=None, timeout=-1, async_=False):
           cmd = "saul"
           if args is not None:
               cmd += " {args}".format(args=" ".join(str(a) for a in args))
           return self.cmd(cmd, timeout=timeout, async_=False)

       def saul_read(self, dev_id="all", timeout=-1, async_=False):
           return self.saul_cmd(args=("read", f"{dev_id}",), timeout=timeout, async_=async_)

       def saul_help(self, timeout=-1, async_=False):
           return self.saul_cmd(args=("help",), timeout=timeout, async_=async_)

> Extending with ``saul_write`` command is left as an exercise.

Reviewer: I've heard people hate this if sentences like these appear in _documentation_ ;-).

Author: ok, removed

> Parsing SAUL Interaction Results

Reviewer: Same then for this :-).

Author: done

   ctrl1 = RIOTCtrl(env=env1, application_directory='.')
   # second samr21-xpro
   env2 = {'BOARD': 'samr21-xpro', 'DEBUG_ADAPTER_ID': 'ATML2127031800011458'}
   ctrl2 = RIOTCtrl(env=env2, application_directory='.')

For the advanced user one could also do as suggested in
`multiple-boards-udev <https://api.riot-os.org/advanced-build-system-tricks.html#multiple-boards-udev>`__
and use an easy to remember variable to identify BOARDs (which would
allow also running the same python code on different setups), if
following the above guide:

.. code:: python

   # first samr21-xpro
   env1 = {'BOARD': 'samr21-xpro', 'BOARD_NUM': 0}
   ctrl1 = RIOTCtrl(env=env1, application_directory='.')
   # second samr21-xpro
   env2 = {'BOARD': 'samr21-xpro', 'BOARD_NUM': 1}
   ctrl2 = RIOTCtrl(env=env2, application_directory='.')

Factories
~~~~~~~~~

The same tasks are done multiple times creating the object flashing it,
starting the terminal and making sure its clean up. Once experiments
grow and take over multiple ctrls this can become tedious, using a
Factory together with a context manager can help with this.

Going back to our example lets write a factory inheriting from

> ``RIOTCtrlFactoryBase``:

Reviewer: That class is not mentioned in the code below.

.. code:: python

   from contextlib import ContextDecorator

>    from riotctrl.ctrl import RIOTCtrl, RIOTCtrlBoardFactory

Reviewer: I think RIOTCtrlBoardFactory is in RIOT's `riotctrl_ctrl`, but I can be mistaking.

Author: Nope, its in `riotctrl.ctrl`: https://github.com/RIOT-OS/riotctrl/blob/ba939f7a39d358f8d29ccf0feae0d57758460e46/riotctrl/ctrl.py#L227-L270

   from riotctrl_ctrl import native

   class RIOTCtrlAppFactory(RIOTCtrlBoardFactory, ContextDecorator):

       def __init__(self):
           super().__init__(board_cls={
               'native': native.NativeRIOTCtrl,
           })

>            self.ctrl_list = list()

Reviewer: ```suggestion
           self.ctrl_list = []
```

Is what I see more often to initialize an empty list.

       def __enter__(self):
           return self

       def __exit__(self, *exc):
           for ctrl in self.ctrl_list:

>                ctrl.stop_term()

Reviewer: IMHO this should be part of `RIOTCtrlFactoryBase` (or a terminal-specific extension of it?). Many will use it as such, when documented here, so let's prevent future copy-paste, and put it into the base code.

Author: IMO that would be out of scope, can we first get a base README in and the do modifications to RIOTCtrl?

       def get_ctrl(self, application_directory='.', env=None):
           # retrieve a RIOTCtrl Object
           ctrl = super().get_ctrl(
               env=env,
               application_directory=application_directory
           )
           # append ctrl to list

>            self.ctrl_list.append(ctrl)

Reviewer: Same for this (+ maybe it should be a set, rather than a list?)

Author: Can we leave this for a code PR?

           # flash and start terminal
           ctrl.flash()
           ctrl.start_term()
           # return ctrl with started terminal
           return ctrl

And the script itself can be re-written as:

.. code:: python

   with RIOTCtrlAppFactory() as factory:
       env = {'BOARD': 'native'}
       ctrl = factory.get_ctrl(env=env)
       shell = SaulShell(ctrl)
       parser = SaulShellCmdParser()
       print(parser.parse(shell.saul_cmd()))

GNRC Networking example native
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Lets put all the above into practice and script an experiment verifying
connectivity between two ctrls, here multiple ``native`` instance will
be used.

First create two tap interfaces:

.. code:: shell

>    sudo ${RIOTBASE}/dist/tools/tapsetup/tapsetup -c 2

Reviewer: Should we mention this script here? What if it changes in the future, then this doc, if we don't pay attention, will be outdated.

Author: Would you suggest omitting the how-to, or inlining for linux?

Author: unrolled

   from riotctrl_shell.netif import Ifconfig

   class Shell(ifconfig, GNRCICMPv6Echo):
     pass

   with RIOTCtrlAppFactory() as factory:
       # Create two native instances, specifying the tap interface
       native_0 = factory.get_ctrl(env={'BOARD':'native', 'PORT':'tap0'})
       native_1 = factory.get_ctrl(env={'BOARD':'native', 'PORT':'tap1'})
       # `NativeRIOTCtrl` allows for `make reset` with `native`
       native_0.reset()
       native_1.reset()
       # Perform a multicast ping and parse results
       pinger = Shell(native_0)
       parser = GNRCICMPv6EchoParser()
       result = parser.parse(pinger.ping6("ff02::1"))
       assert result['stats']['packet_loss'] < 10    # assert packetloss is under 10%
       assert result['stats']['rx'] > 0              # assert at least one responder

A more complex example can be seen in the Release Tests:
`04-single-hop-6lowpan-icmp <https://github.com/RIOT-OS/Release-Specs/blob/master/04-single-hop-6lowpan-icmp/test_spec04.py>`__

Examples
~~~~~~~~

-  pytest: `ReleaseSpecs <https://github.com/RIOT-OS/Release-Specs>`__
-  unittests:
   `tests/turo <https://github.com/RIOT-OS/RIOT/blob/master/tests/turo/tests/01-run.py>`__,

>    `tests/congure_test <https://github.com/RIOT-OS/RIOT/blob/master/tests/congure_test/tests/01-run.py>`__

Reviewer: Didn't we agree during the breakout session to also add the discussion section in one way or another? Something like "Future extensions": CoAP/CBOR, Security checks, ...

Author: Added part of the discussion section
